app/src/main/java/org/kiwix/kiwixmobile/main/AddNoteDialog.java

```diff
@@ -149,6 +158,60 @@ public void afterTextChanged(Editable s) {}
     return view;
   }
 
+  private String getZimNoteDirectoryName() {
+    String zimFileName = ZimContentProvider.getZimFile(); // Returns name of the form ".../Kiwix/granbluefantasy_en_all_all_nopic_2018-10.zim"
+
+    String noteDirectoryName = getTextAfterLastSlashWithoutExtension(zimFileName);
+
+    if(noteDirectoryName != null) {
+      return noteDirectoryName;
+    } else {
+      return (zimFileTitle); // Incase the required ZIM file name couldn't be extracted
+    }
+  }
+
+  private String getArticleNotefileName() {
+    String articleUrl = ((MainActivity) getActivity()).getCurrentWebView().getUrl(); // Returns url of the form: "content://org.kiwix.kiwixmobile.zim.base/A/Main_Page.html"
+
+    String notefileName = getTextAfterLastSlashWithoutExtension(articleUrl);
+
+    if(notefileName != null) {
+      return notefileName;
+    } else {
+      return (articleTitle); // Incase the required html file name couldn't be extracted
+    }
+  }
+
+  private String getTextAfterLastSlashWithoutExtension(String path) {
```

Comment:
You shouldn't go introducing nulls into code if you can avoid it, in java we have Nullable and NotNullable annotations to mark methods and Kotlin has Optionals in it's type system.
It is often better to return something empty because that can't crash

Reply:
[Done](https://github.com/kiwix/kiwix-android/pull/1237/commits/085ed7f7bd8c0916f1d1a3ad25a705a3bf1bdfd4#diff-36e13fa4a9132616aee81e1f88bfa0ffR192)

I had a query: Is it ever a bad idea to completely annotate the methods that I'm writing myself (like [so](https://github.com/kiwix/kiwix-android/pull/1237/commits/085ed7f7bd8c0916f1d1a3ad25a705a3bf1bdfd4#diff-36e13fa4a9132616aee81e1f88bfa0ffR177))? Not adding them before feels like a glaring omission in all the code I have written previously, and as you pointed out will definitely limit crashes

Reply:
It is definitely not a bad idea but it just takes forever in java, this is why it should be in the type system. Java introduced Optionals via Guava and they saw a good bit of use but now kotlin is here and the only reason you would annotate java code now is if you are not allowed rewrite it in kotlin.
Lots of effort is put in to AOSP to add these annotations.

Reply:
I'll keep adding them to my Java code from now, thanks